Propagate early characteristic changes

twyatt opened the pull request:
Closes #105
Closes #112

Also fixes JavaScript observations to wait until connection is ready. Only users of experimental advertisement support in Chrome would've seen this bug, as using `requestDevice` obtains devices in a connected state, i.e. already ready.

sdonn3 commented on core/src/androidMain/kotlin/Observers.kt:
```diff
@@ -38,20 +40,21 @@ internal class Observers(
-                    it.characteristic.serviceUuid == characteristic.serviceUuid
-                ) emit(it.data)
+    fun acquire(
+        characteristic: Characteristic
```
nit: seems like a single parameter could just go on one line?

sdonn3 reviewed (approved):
looks good!

twyatt commented:
@sdonn3 @cedrickcooke re-requested review as I've added a additional changes since last reviewed:

- refactored JavaScript observation handling to resemble the other platforms
- implemented `observe`'s `onSubscription` lambda to be invoked on re-connection
    - now supports invoking all `onSubscription` lambdas when multiple `observe` Flows are created for the same `characteristic`

cedrickcooke commented on core/src/commonMain/kotlin/Peripheral.kt:
```diff
@@ -114,8 +115,27 @@ public interface Peripheral {
      * then notification failures are propagated via the returned [observe] [Flow].
      *
      * If the specified [characteristic] is invalid or cannot be found then a [NoSuchElementException] is propagated.
+     *
+     * The optional [onSubscription] parameter is functionally identical to using the
+     * [onSubscription][kotlinx.coroutines.flow.onSubscription] operator on the returned [Flow] except it has the
+     * following special properties:
+     *
+     * - It will be executed whenever [connection][connect] is established (while the returned [Flow] is active); and
+     * - It will be executed _after_ the observation is spun up (i.e. after enabling notifications or indications)
+     *
+     * The [onSubscription] action is useful in situations where an initial operation is needed when starting an
+     * observation (such as writing a configuration to the peripheral and expecting the response to come back in the
+     * form of a characteristic change). The [onSubscription] is invoked for every new subscriber; if it is desirable to
+     * only invoke the [onSubscription] once per connection (for the specified [characteristic]) then you can either
+     * use the [shareIn][kotlinx.coroutines.flow.shareIn] [Flow] operator on the returned [Flow], or call [observe]
+     * again with the same [characteristic] and without specifying an [onSubscription] action.
+     *
+     * If multiple [observations][observe] are created for the same [characteristic] but with different [onSubscription]
+     * actions, then the [onSubscription] actions will be executed in the order in which the returned [Flow]s are
+     * subscribed to.
```
❤️

cedrickcooke commented on core/src/jsMain/kotlin/Observers.kt:
```diff
@@ -1,129 +1,89 @@
-    fun invalidate() {
-        observers.forEach { (_, observation) ->
-            observation.listener = null
+private class Observations {
```
It's a little sad that this `Observations` class can't live in common code. I can see that they're different but it feels _super_ similar between the three implementations.

twyatt replied:
I agree it's a little sad. Maybe one day I can figure out a way to make it generic and reusable, but for now, just getting it working. 🤷

twyatt commented:
Going to do some internal testing against the JavaScript target, then merge.

sdonn3 commented on core/src/androidMain/kotlin/Observers.kt:
```diff
@@ -34,25 +36,26 @@ internal class Observers(
+    ) = characteristicChanges
+        .onSubscription {
+            peripheral.suspendUntilReady()
+            if (observations.add(characteristic, onSubscription) == 1) {
```
Should this be '>= 1' to mirror the '< 1' below?

sdonn3 reviewed (approved):
Totally appreciate how easy this is to read through

twyatt commented:
lol, @sdonn3 either I'm going crazy, or you left a comment then deleted it.

Here was the response I was going to do, though I'm guessing you don't need it now. 🤷 

----

Technically, the `< 1` below could be a `0` (but `< 1` was a little bit of a safety buffer), whereas this can only be a `== 1`.

I'll explain why:

In `onSubscription` we "register", aka "add", a subscription to the `Map` of characteristics to subscriptions. The return is the number of subscribers we have _after_ "registering" (or aka "adding"). The underlying system only needs to setup BLE notifications/indications once per characteristic (and not more than once). So, we essentially only want to enable them when we go from `0` to `1` subscribers and we'll just leave the notifications/indications enabled as more subscribers jump on.

On the reverse side (in `onCompletion`) we need to keep track as `Flow` collectors unsubscribe, all the way down until we have no more subscribers, and can therefore stop the BLE notifications/indications for that characteristic. Since we're subtracting within the `observations.remove` I thought it were safer if I messed up the math (and we somehow ended up with `-1`) that checking for `< 1` would be good. But, assuming the maths checks out 😉  then `== 0` would've sufficed. Essentially it would've represented when we are transitioning from `1` to `0` subscribers.
